Add in-game option to switch fullscreen monitor

Francessco121 opened the pull request:
Worth mentioning that the selected location is referred to as 'monitor' in code but 'display' on the UI. Monitor is consistent with GLFW's terms, but display makes more sense to users. The options are also listed as just 'DISPLAY 1', 'DISPLAY 2', etc. I would have preferred to show the monitor's name but I wasn't sure how to get those strings into GOAL.

Let me know if I went a little overboard on the refactoring, I ran into some issues tracking the previous windowed pos/size when going from fullscreen to windowed, so I changed it quite a bit.  

Also, I had to modify the ImGui source, their GLFW implementation hijacks the monitor changed callback but never calls our registered callback (unlike the other events they steal). I have no idea why that code is even in ImGui's current release since their callback is empty and marked as such...

Closes #1454

water111 commented on third-party/imgui/imgui_impl_glfw.cpp:
```diff
@@ -197,8 +197,12 @@ void ImGui_ImplGlfw_CharCallback(GLFWwindow* window, unsigned int c)
+  ImGui_ImplGlfw_Data* bd = ImGui_ImplGlfw_GetBackendData();
+  if (bd->PrevUserCallbackMonitor != NULL)
+    bd->PrevUserCallbackMonitor(monitor, event);
+
```
wow, was imgui just missing this originally?

Francessco121 replied:
Yep, looking at the current release you can find the same empty function here https://github.com/ocornut/imgui/blob/v1.88/backends/imgui_impl_glfw.cpp#L397. Absolutely no idea why it's like this.

water111 reviewed (commented):
Sorry, I totally missed this PR!  If I forget about stuff like this in the future feel free to ping me on discord or in the PR.

Unfortunately I'm away this weekend so I can't test right now, so if somebody else with multiple monitors could give it a try, that would be helpful.

Some things that I'd recommend trying:
- try the "reboot in debug mode" option a few times, just to make sure we can still recreate the window. (the logic here all looks good, but it's easy to have bugs here)
- make sure we aren't adding too many more glfw function calls per frame.  At least on my computer they take 0.1ms each so we spent a pretty significant amount of time doing these. As far as I can tell this PR doesn't add any.
- unplug/connect monitor as the game is running. make sure that glfw window size/pos callbacks happen if the OS moves the window off of a monitor that was just disconnected. You can get a similar effect by changing the resolution (at least in windows, this forces the window positions to all be reassigned)

Francessco121 commented:
No worries! I've been a bit busy myself. I did test the debug mode switch and unplugging monitors while the game was fullscreen on the monitor that got unplugged and such (actually ran into several issues with both tests originally), so it should be fine. It's been a bit tho so I'll double check.

For the additional glfw calls, I tried to avoid making any during each frame using the new glfw callbacks and just keeping track of the state. I'll double check this as well tho.

ManDude reviewed (changes requested):
I only have 1 monitor, so I can't do thorough testing, but it seems to work.

You need to add your new menu to the `can-go-back?` function so users can press start or square to exit the menu.

Francessco121 commented on goal_src/jak1/engine/ui/progress/progress-static.gc:
```diff
@@ -151,7 +151,7 @@
 
 ;; maps options to a progress screen
-(define *options-remap* (new 'static 'boxed-array :type (array game-option) :length 0 :allocated-length (#if (not PC_PORT) 35 60)))
+(define *options-remap* (new 'static 'boxed-array :type (array game-option) :length 0 :allocated-length (#if (not PC_PORT) 35 61)))
```
Yep, turns out 60 is indeed enough.

Francessco121 commented:
Thanks for the feedback, I've addressed each comment.

I changed up how the fullscreen and minimized states are retrieved, the GLDisplay class now tracks them as fields so their states can be retrieved without calling GLFW. Turns out that this PR did originally add an extra GLFW call per frame, doing this removes that. However, I wasn't able to fully remove all GLFW calls used each frame to check if the fullscreen mode needs to change. Unfortunately, GLFW doesn't have a callback for video mode changes and without that or checking it each frame the game has no idea if you change your monitor's resolution for example. At least on Windows however, this call only takes ~0.01ms.

I also did the following tests:
- Clicking 'restart in debug mode' multiple times in each fullscreen mode on different monitors.
- Unplugging the monitor that the game is fullscreened (or borderless) on (game moves over to the primary monitor).
- Changing the resolution of the monitor that the game is fullscreened (or borderless) on.
    - Note: The resolution that the game renders at doesn't get affected by this, but GLFW will be given the correct video mode when this changes so that the game understands the active monitor's dimensions.
- Unplugging the monitor that the game is fullscreened on and forcing it to default to a different monitor with a different resolution.
- Same unplugging tests but with the Windows monitor 'Project' feature.

I think everything should be all set. Please let me know if I committed any C++ sins tho, I'd be happy to change any of that.

xTVaser commented:
add to whitelist
